test/CMakeLists.txt

```diff
@@ -7,72 +7,163 @@ FetchContent_Declare(doctest
 )
 FetchContent_MakeAvailable(doctest)
 
-add_library(sfml-test-main STATIC
-    DoctestMain.cpp
-    TestUtilities/SystemUtil.hpp
-    TestUtilities/SystemUtil.cpp
-    TestUtilities/WindowUtil.hpp
-    TestUtilities/WindowUtil.cpp
-    TestUtilities/GraphicsUtil.hpp
-    TestUtilities/GraphicsUtil.cpp
-)
-target_include_directories(sfml-test-main PUBLIC TestUtilities)
-target_compile_definitions(sfml-test-main PUBLIC DOCTEST_CONFIG_REQUIRE_STRINGIFICATION_FOR_ALL_USED_TYPES)
-target_link_libraries(sfml-test-main PUBLIC SFML::System doctest::doctest)
-set_target_warnings(sfml-test-main)
-
-SET(SYSTEM_SRC
-    System/Angle.cpp
-    System/Clock.cpp
-    System/Err.cpp
-    System/FileInputStream.cpp
-    System/MemoryInputStream.cpp
-    System/Time.cpp
-    System/Vector2.cpp
-    System/Vector3.cpp
-)
-sfml_add_test(test-sfml-system "${SYSTEM_SRC}" SFML::System)
+# add a new static library which provides a main function and testing utilities for SFML types
+# example: sfml_add_test_utils(sfml-window-test-utils
+#                              SOURCES TestUtilities/WindowUtil.hpp TestUtilities/WindowUtil.cpp
+#                              DEPENDS SFML::Window sfml-system-test-utils)
+function(sfml_add_test_utils target)
+
+    # parse the arguments
+    cmake_parse_arguments(THIS "" "" "SOURCES;DEPENDS" ${ARGN})
+    if (NOT "${THIS_UNPARSED_ARGUMENTS}" STREQUAL "")
+        message(FATAL_ERROR "Extra unparsed arguments when calling sfml_add_test_utils: ${THIS_UNPARSED_ARGUMENTS}")
+    endif()
+
+    # set a source group for the source files
+    source_group("" FILES ${THIS_SOURCES})
+
+    # create the target
+    add_library(${target} STATIC ${THIS_SOURCES})
+
+    # set the target's folder (for IDEs that support it, e.g. Visual Studio)
+    set_target_properties(${target} PROPERTIES FOLDER "Tests/Utilities")
+
+    # link the target to its dependencies
+    target_link_libraries(${target} PUBLIC ${THIS_DEPENDS})
+
+    set_target_warnings(${target})
+endfunction()
+
+# add a new target which is a test executable for an SFML module
+# example: sfml_add_test(Network
+#                        SOURCES ftp.cpp ...
+#                        DEPENDS SFML::Network sfml-system-test-utils)
+function(sfml_add_test module)
+
+    # parse the arguments
+    cmake_parse_arguments(THIS "" "" "SOURCES;DEPENDS" ${ARGN})
+    if (NOT "${THIS_UNPARSED_ARGUMENTS}" STREQUAL "")
+        message(FATAL_ERROR "Extra unparsed arguments when calling sfml_add_test: ${THIS_UNPARSED_ARGUMENTS}")
+    endif()
+
+    # set a source group for the source files
+    source_group("" FILES ${THIS_SOURCES})
+
+    # create the target
+    string(TOLOWER test-sfml-${module} target)
```

Comment:
What do you think about `sfml_add_test_utils` also using this same naming logic where instead of providing a target name, you just provide the module name?

Reply:
I was tempted to do it, but I think it might make the code a little less clear because it would be less obvious where the test utility library target names come from. But I don't feel strongly about it, I could be convinced otherwise.

Consider:
```cmake
sfml_add_test_utils(sfml-system-test-utils ...)
target_include_directories(sfml-system-test-utils PUBLIC TestUtilities)
```
versus
```cmake
sfml_add_test_utils(System ...)
target_include_directories(sfml-system-test-utils PUBLIC TestUtilities)
```